### DISCUSSION TITLE
how to convert proto file from text to binary form ?

### DISCUSSION THREAD

**Discussion by gregy4** [google/or-tools#2460] [CP-SAT questions] [answered]
Hello,
  how can I convert proto file in text format to binary for cp-sat problem ?

Thanks,
Jan

**Comment by lperron** [comment, marked as answer]
why ?

Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by gregy4** [reply]
I want to proto file created after presolve to load and modify, this is in text format. In java I can load proto file but only binary form is accepted via mergeFrom method. Maybe some alternative via TextFormat like in python can work but I didn’t try it so far and I’m not sure whether it works.

**Reply by lperron** [reply]
I guess using https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/TextFormat to read the proto first.

**Reply by gregy4** [reply]
It finally works with TextFormat, protobuf in java and python use for unknown reason different method names. This was main problem for me. There's also question why is not present in builder alternative method to load text proto file.

java:
```java
CpModel model = new CpModel();
Reader inputReader = new InputStreamReader(new FileInputStream(filename));
TextFormat.merge(inputReader, model.getBuilder());
```

python:
```python
model = cp_model.CpModel()
text_format.Parse(protoString, model.Proto())
```

**Comment by lperron** [comment]
great!
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00
